src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp

```diff
@@ -0,0 +1,186 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+
+#include <array>
+#include <cstddef>
+#include <functional>
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Transpose.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+void apply_matrix_in_first_dim(double* result, const double* const input,
```

Comment:
Is there any reason not to hide the pointers instead this function? E.g. pass in DataVectors and call .data() in this function? I guess it doesn't matter much, since the function is only used internally here, but in general, I would have thought it would be nicer to have an interface that avoids passing around the raw pointers, unless there's some performance penalty

Reply:
I agree DataVectors is nicer. The reason I used `double*` is to be consistent with `ApplyMatrices.cpp` and `WeakDivergence.hpp`. Since #2696 should remove/simplify most of this code I would prefer to keep it consistent to make replacement more obvious.